File: exercises/cookbook/attributes.cpp
// Appendix F, Recipe 51 - tell the compiler what a function promises.
//
// session_status(), describe(), fatal() and retire_session() are included by
// book/F-rosetta-cookbook.md between the recipe-51 markers: edit here and the
// page follows. main() is scaffolding.
//
// An attribute changes no instruction, so a running program cannot check one.
// What each attribute changes is a DIAGNOSTIC, and the judge is therefore
// three builds that must be REFUSED - Appendix I's constlab discipline,
// applied to warnings instead of to const. scripts/build_all.sh compiles this
// file three more times with -Werror and one -D each, and asserts that the
// compiler says no and says why:
//
//   -DATTR_DISCARD_RESULT   ignores a [[nodiscard]] result   -> "nodiscard"
//   -DATTR_NO_NORETURN      removes [[noreturn]] from fatal  -> "return-type"
//   -DATTR_USE_DEPRECATED   calls the [[deprecated]] one     -> "deprecated"
//
// Take the attribute away and the clean build stays clean, which is exactly
// why the refusals are the check: nothing else here would notice.
#include <cassert>
#include <cstdio>
#include <cstdlib>
#include <string>

#ifdef ATTR_NO_NORETURN
#  define MAYBE_NORETURN
#else
#  define MAYBE_NORETURN [[noreturn]]
#endif

enum class Status { Ok, Busy, Failed };

// --8<-- [start:recipe-51]
// Never comes back. The caller needs no `return` after it, and the compiler
// stops asking about the path that falls off the end.
MAYBE_NORETURN void fatal(const char* why) {
    std::fprintf(stderr, "fatal: %s\n", why);
    std::abort();
}

// The result is the whole point of the call: ignoring it is a warning, and a
// build with -Werror refuses it. This is the compiler enforcing the sentence
// Chapter 8 spends a page on - a failure that is a value has to be looked at.
[[nodiscard]] Status session_status(int id) {
    return id > 0 ? Status::Ok : Status::Failed;
}

const char* describe(Status status, [[maybe_unused]] int verbosity) {
    switch (status) {
        case Status::Ok:
            return "ok";
        case Status::Busy:
            [[fallthrough]];          // deliberate, and said out loud
        case Status::Failed:
            return "not available";
    }
    return "unknown";
}

// Where [[noreturn]] earns its keep: a function with nothing to return on the
// failing path, because the caller's contract is broken and there is no
// sensible value (Chapter 8's assert row, in a codebase that ships with
// NDEBUG). There is no `return` after the call, and
// with the attribute the compiler needs none; without it, it warns that
// control reaches the end of a non-void function - which is the whole
// difference an attribute makes, and the reason one of the refusals below
// removes it.
int required_channel(int configured) {
    if (configured > 0) {
        return configured;
    }
    fatal("channel not configured - the caller's contract, broken");
}

// Still compiles, still links, still ships; every call site gets a warning
// naming the replacement. C#'s [Obsolete], with the message in the same place.
[[deprecated("use session_status")]] Status retire_session(int id) {
    return session_status(id);
}
// --8<-- [end:recipe-51]

int main() {
    // The values, which the attributes do not touch - stated so the listing
    // is a program and not five declarations.
    const Status ok = session_status(1);
    const Status bad = session_status(0);
    assert(ok == Status::Ok);
    assert(bad == Status::Failed);
    assert(std::string(describe(ok, 0)) == "ok");
    assert(std::string(describe(bad, 0)) == "not available");
    assert(std::string(describe(Status::Busy, 0)) == "not available");
    assert(required_channel(3) == 3);

#ifdef ATTR_DISCARD_RESULT
    // Must be refused: the result is [[nodiscard]] and nobody looks at it.
    session_status(1);
#endif
#ifdef ATTR_USE_DEPRECATED
    // Must be refused: the function still exists, and saying so is the point.
    (void)retire_session(1);
#endif
    return 0;
}

File: exercises/cookbook/rust/src/attributes.rs
//! Appendix F, Recipe 51 - tell the compiler what a function promises.
//!
//! The C++ side needs three builds that must fail, because an attribute's
//! whole effect is on a diagnostic. Rust's answers are the same three ideas
//! with the enforcement moved: `#[must_use]` is `[[nodiscard]]`, the `!`
//! return type is `[[noreturn]]` written in the type system rather than
//! beside it, and `#[deprecated]` is the same attribute under the same name.

// --8<-- [start:recipe-51]
pub enum Status {
    Ok,
    Busy,
    Failed,
}

// `!` is the never type: this function has no return value because it does
// not return, and that is a fact about its TYPE, so the caller needs no
// `return` after it and the compiler needs no attribute to be told.
pub fn fatal(why: &str) -> ! {
    panic!("fatal: {why}");
}

// `#[must_use]` is `[[nodiscard]]`, and the lint is on by default.
#[must_use]
pub fn session_status(id: i32) -> Status {
    if id > 0 { Status::Ok } else { Status::Failed }
}

pub fn describe(status: &Status, _verbosity: i32) -> &'static str {
    // No fallthrough to say out loud: arms do not fall through, and the
    // `|` pattern is how two of them share a body. The match is exhaustive
    // or it does not compile, which is the other half of `[[fallthrough]]`'s
    // job done by the language.
    match status {
        Status::Ok => "ok",
        Status::Busy | Status::Failed => "not available",
    }
}

pub fn required_channel(configured: i32) -> i32 {
    if configured > 0 {
        return configured;
    }
    fatal("channel not configured - the caller's contract, broken")
}

#[deprecated(note = "use session_status")]
pub fn retire_session(id: i32) -> Status {
    session_status(id)
}
// --8<-- [end:recipe-51]

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn the_values_the_attributes_do_not_change() {
        assert!(matches!(session_status(1), Status::Ok));
        assert!(matches!(session_status(0), Status::Failed));
        assert_eq!(describe(&Status::Ok, 0), "ok");
        assert_eq!(describe(&Status::Busy, 0), "not available");
        assert_eq!(required_channel(3), 3);
    }

    #[test]
    #[should_panic(expected = "channel not configured")]
    fn the_never_type_really_does_not_return() {
        required_channel(0);
    }
}
